File: client/src/bastionlab/linfa/trainers.py
from dataclasses import dataclass, field
from ..pb.bastionlab_linfa_pb2 import (
    LinearRegression as ProtoLinReg,
    LogisticRegression as ProtoLogReg,
    KMeans as ProtoKMeans,
    DecisionTree as ProtoDecisionTree,
    ElasticNet as ProtoElas,
    GaussianNb as ProtoGaussian,
    SVM,
    _SVM_KERNELPARAMS_N as N,
)
from ..polars.remote_polars import RemoteArray
from typing import Dict, Optional, List
from enum import Enum

@dataclass
class Trainer:
    def to_msg_dict():
        raise NotImplementedError

    def fit(self, train_set: "RemoteArray", target_set: "RemoteArray"):
        client = train_set._client.linfa
        model = client._train(train_set, target_set, self)
        self.identifier = model.identifier
        return model

    def predict(self, test_set: "RemoteArray"):
        client = test_set._client.linfa
        return client._predict(self, test_set)


@dataclass
class GaussianNb(Trainer):
    var_smoothing: float = 1e-9

    def to_msg_dict(self):
        return {"gaussian_nb": ProtoGaussian(var_smoothing=self.var_smoothing)}

    def fit(self, train_set: "RemoteArray", target_set: "RemoteArray"):
        return super().fit(train_set, target_set)

    def predict(self, test_set: "RemoteArray"):
        return super().predict(test_set)


@dataclass
class LinearRegression(Trainer):
    fit_intercept: bool = True

    def to_msg_dict(self):
        return {"linear_regression": ProtoLinReg(fit_intercept=self.fit_intercept)}

    def fit(self, train_set: "RemoteArray", target_set: "RemoteArray"):
        return super().fit(train_set, target_set)

    def predict(self, test_set: "RemoteArray"):
        return super().predict(test_set)


@dataclass
class LogisticRegression(Trainer):
    alpha: float = 1.0
    gradient_tolerance: float = 1e-4
    fit_intercept: bool = True
    max_iterations: int = 100
    initial_params: List[float] = field(default_factory=list)

    def to_msg_dict(self):
        return {
            "logistic_regression": ProtoLogReg(
                alpha=self.alpha,
                gradient_tolerance=self.gradient_tolerance,
                fit_intercept=self.fit_intercept,
                max_iterations=self.max_iterations,
                initial_params=self.initial_params,
            )
        }

    def fit(self, train_set: "RemoteArray", target_set: "RemoteArray"):
        return super().fit(train_set, target_set)

    def predict(self, test_set: "RemoteArray"):
        return super().predict(test_set)


@dataclass
class ElasticNet(Trainer):
    penalty: float = 0.1
    l1_ratio: float = 0.1
    with_intercept: bool = False
    max_iterations: int = 1000
    tolerance: float = 1e-4

    def to_msg_dict(self):
        return {
            "elastic_net": ProtoElas(
                penalty=self.penalty,
                l1_ratio=self.l1_ratio,
                with_intercept=self.with_intercept,
                max_iterations=self.max_iterations,
                tolerance=self.tolerance,
            )
        }

    def fit(self, train_set: "RemoteArray", target_set: "RemoteArray"):
        return super().fit(train_set, target_set)

    def predict(self, test_set: "RemoteArray"):
        return super().predict(test_set)


@dataclass
class DecisionTree(Trainer):
    class SplitQuality(Trainer):
        def to_msg_dict():
            pass

    class Gini(SplitQuality):
        def to_msg_dict():
            return {"gini": ProtoDecisionTree.Gini}

    class Entropy(SplitQuality):
        def to_msg_dict():
            return {"entropy": ProtoDecisionTree.Entropy}

    split_quality: Optional[SplitQuality] = Gini
    max_depth: Optional[int] = None
    min_weight_split: float = (2.0,)
    min_weight_leaf: float = (1.0,)
    min_impurity_decrease: float = 0.00001

    def to_msg_dict(self):
        return {
            "decision_tree": ProtoDecisionTree(
                max_depth=self.max_depth,
                min_weight_split=self.min_weight_split,
                min_weight_leaf=self.min_weight_leaf,
                min_impurity_decrease=self.min_impurity_decrease,
                split_quality=self.split_quality.to_msg_dict(),
            )
        }

    def fit(self, train_set: "RemoteArray", target_set: "RemoteArray"):
        return super().fit(train_set, target_set)

    def predict(self, test_set: "RemoteArray"):
        return super().predict(test_set)


@dataclass
class KMeans(Trainer):
    class InitMethod(Trainer):
        def to_msg_dict():
            pass

    class Random(InitMethod):
        def to_msg_dict():
            return {"random": ProtoKMeans.Random()}

    class KMeanPara(InitMethod):
        def to_msg_dict():
            return {"kmeans_para": ProtoKMeans.KMeansPara()}

    class KMeansPlusPlus(InitMethod):
        def to_msg_dict():
            return {"kmeans_plus_plus": ProtoKMeans.KMeansPlusPlus()}

    n_runs: int = 10
    n_clusters: int = 0
    tolerance: float = 1e-4
    max_n_iterations: int = 300
    init_method: Optional[InitMethod] = None

    def get_init_method(self) -> Dict:
        return (
            self.Random.to_msg_dict()
            if not self.init_method
            else self.init_method.to_msg_dict()
        )

    def to_msg_dict(self):
        return {
            "kmeans": ProtoKMeans(
                n_runs=self.n_runs,
                n_clusters=self.n_clusters,
                tolerance=self.tolerance,
                max_n_iterations=self.max_n_iterations,
                **self.get_init_method(),
            )
        }

    def fit(self, train_set: "RemoteArray", target_set: "RemoteArray"):
        return super().fit(train_set, target_set)

    def predict(self, test_set: "RemoteArray"):
        return super().predict(test_set)


@dataclass
class SVC(Trainer):
    class PlattParams(Trainer):
        maxiter: int = 100
        ministep: float = 1e-10
        sigma: float = 1e-12

        def to_msg_dict(self):
            return {
                "platt_params": SVM.PlattParams(
                    maxiter=self.maxiter, ministep=self.ministep, sigma=self.sigma
                )
            }

    class KernelParams(Trainer):
        class KernelMethod:
            pass

        class KernelType(Trainer):
            pass

        class Dense(KernelType):
            def to_msg_dict():
                return {"dense": SVM.KernelParams.Dense()}

        class Sparse(KernelType):
            sparsity: int = 1

            def to_msg_dict(self):
                return {"sparse": SVM.KernelParams.Sparse(sparsity=self.sparsity)}

        class Gaussian(KernelMethod):
            eps: float = 1.0

            def to_msg_dict(self):
                return {"gaussian": SVM.KernelParams.Guassian(eps=self.eps)}

        class Linear(KernelMethod):
            def to_msg_dict():
                return {"linear": SVM.KernelParams.Linear()}

        class Polynomial(KernelMethod):
            constant: float = 1.0
            degree: float = 1.0

            def to_msg_dict(self):
                return {
                    "poly": SVM.KernelParams.Polynomial(
                        constant=self.constant, degree=self.degree
                    )
                }

        kernel_method: Optional[KernelMethod] = Linear.to_msg_dict()
        kernel_type: Optional[KernelType] = Dense.to_msg_dict()
        n: N = SVM.KernelParams.LinearSearch

        def to_msg_dict(self):
            return {
                "kernel_params": SVM.KernelParams(
                    **self.kernel_method, **self.kernel_type, n=self.n
                )
            }

    c: List[float] = field(default_factory=list)
    eps: Optional[float] = None
    nu: Optional[float] = None
    shrinking: bool = False
    platt_params: Optional[PlattParams] = None
    kernel_params: Optional[KernelParams] = None

    def get_kernel_params(self):
        return (
            self.KernelParams().to_msg_dict()
            if not self.kernel_params
            else self.kernel_params.to_msg_dict()
        )

    def get_platt_params(self):
        return (
            self.PlattParams().to_msg_dict()
            if not self.platt_params
            else self.platt_params.to_msg_dict()
        )

    def fit(self, train_set: "RemoteArray", target_set: "RemoteArray"):
        return super().fit(train_set, target_set)

    def predict(self, test_set: "RemoteArray"):
        return super().predict(test_set)

    def to_msg_dict(self):
        return {
            "svm": SVM(
                c=self.c,
                eps=self.eps,
                shrinking=self.shrinking,
                nu=self.nu,
                **self.get_kernel_params(),
                **self.get_platt_params(),
            )
        }

File: server/bastionlab_linfa/src/trainers.rs
use bastionlab_common::array_store::ArrayStore;
use linfa::{prelude::Records, DatasetBase, PlattParams};
use linfa_bayes::GaussianNb;
use linfa_clustering::{KMeans, KMeansInit};
use linfa_elasticnet::ElasticNet;
use linfa_kernel::{Kernel, KernelMethod, KernelParams, KernelType};
use linfa_linear::{FittedLinearRegression, Link, TweedieRegressor};
use linfa_logistic::{FittedLogisticRegression, MultiFittedLogisticRegression};
use linfa_nn::{
    distance::L2Dist, BallTreeIndex, CommonNearestNeighbour, KdTreeIndex, LinearSearchIndex,
};
use linfa_svm::Svm;
use linfa_trees::{DecisionTree, SplitQuality};
use ndarray::{Array2, ArrayBase, Ix1, Ix2, OwnedRepr};

use tonic::{Request, Status};

use crate::{
    linfa_proto::{
        tweedie_regressor::Link as ProtoTweedieLink, ElasticNet as BastionElasticNet,
        KMeans as BastionKMeans, Svm as BastionSvm, Trainer,
        TweedieRegressor as BastionTweedieRegressor, *,
    },
    to_status_error,
};

#[derive(Debug)]
pub enum Models {
    GaussianNaiveBayes {
        var_smoothing: f64,
    },
    ElasticNet {
        penalty: f32,
        l1_ratio: f32,
        with_intercept: bool,
        max_iterations: u32,
        tolerance: f32,
    },
    KMeans {
        n_runs: usize,
        n_clusters: usize,
        tolerance: f64,
        max_n_iterations: u64,
        init_method: KMeansInit<f64>,
    },

    LinearRegression {
        fit_intercept: bool,
    },
    TweedieRegressor {
        fit_intercept: bool,
        alpha: f64,
        max_iter: u64,
        link: Link,
        tol: f64,
        power: f64,
    },
    BinomialLogisticRegression {
        alpha: f64,
        gradient_tolerance: f64,
        fit_intercept: bool,
        max_iterations: u64,
        initial_params: Option<Vec<f64>>,
    },
    MultinomialLogisticRegression {
        alpha: f64,
        gradient_tolerance: f64,
        fit_intercept: bool,
        max_iterations: u64,
        initial_params: Option<Vec<f64>>,
        shape: Option<Vec<u64>>,
    },

    DecisionTree {
        split_quality: SplitQuality,
        max_depth: Option<usize>,
        min_weight_split: f32,
        min_weight_leaf: f32,
        min_impurity_decrease: f64,
    },
    SVM {
        c: Vec<f32>,
        eps: Option<f32>,
        nu: Option<f32>,
        shrinking: bool,
        platt_params: PlattParams<f64, ()>,
        kernel_params: KernelParams<f64>,
    },
}

#[allow(unused)]
#[derive(Debug)]
pub enum SupportedModels {
    GaussianNaiveBayes(GaussianNb<f64, usize>),
    ElasticNet(ElasticNet<f64>),
    KMeans(KMeans<f64, L2Dist>),
    LinearRegression(FittedLinearRegression<f64>),
    TweedieRegressor(TweedieRegressor<f64>),
    BinomialLogisticRegression(FittedLogisticRegression<f64, u64>),
    MultinomialLogisticRegression(MultiFittedLogisticRegression<f64, u64>),
    DecisionTree(DecisionTree<f64, usize>),
    BallTree(BallTreeIndex<'static, f64, L2Dist>),
    Linear(LinearSearchIndex<'static, f64, L2Dist>),
    KdTree(KdTreeIndex<'static, f64, L2Dist>),
    SVM(Svm<f64, f64>),
}

#[derive(Debug)]
pub enum PredictionTypes {
    Usize(DatasetBase<ArrayBase<OwnedRepr<f64>, Ix2>, ArrayBase<OwnedRepr<usize>, Ix1>>),
    U64(DatasetBase<ArrayBase<OwnedRepr<f64>, Ix2>, ArrayBase<OwnedRepr<u64>, Ix1>>),
    Float(DatasetBase<ArrayBase<OwnedRepr<f64>, Ix2>, ArrayBase<OwnedRepr<f64>, Ix1>>),
    SingleProbability(ArrayBase<OwnedRepr<f64>, Ix1>),
    MultiProbability(ArrayBase<OwnedRepr<f64>, Ix2>),
}

#[derive(Debug)]
pub struct IArrayStore(pub ArrayStore);

impl Records for IArrayStore {
    type Elem = IArrayStore;

    fn nsamples(&self) -> usize {
        self.0.height()
    }

    fn nfeatures(&self) -> usize {
        self.0.width()
    }
}

pub fn get_datasets(
    records: ArrayStore,
    target: ArrayStore,
) -> DatasetBase<IArrayStore, IArrayStore> {
    // ** For now, shuffling is not implemented.
    let dataset = DatasetBase::new(IArrayStore(records), IArrayStore(target));
    dataset
}

pub fn process_trainer_req(
    request: Request<TrainingRequest>,
) -> Result<(String, String, Option<Trainer>), Status> {
    Ok((
        request.get_ref().records.clone(),
        request.get_ref().target.clone(),
        Some(
            request
                .get_ref()
                .trainer
                .clone()
                .ok_or_else(|| Status::aborted("Trainer not supported!"))?,
        ),
    ))
}

pub fn select_trainer(trainer: Trainer) -> Result<Models, Status> {
    if let Some(t) = trainer.gaussian_nb {
        Ok(Models::GaussianNaiveBayes {
            var_smoothing: t.var_smoothing.into(),
        })
    } else if let Some(t) = trainer.linear_regression {
        Ok(Models::LinearRegression {
            fit_intercept: t.fit_intercept,
        })
    } else if let Some(t) = trainer.binomial_logistic_regression {
        let initial_params = t
            .initial_params
            .iter()
            .map(|v| (*v).into())
            .collect::<Vec<f64>>();

        let initial_params = if initial_params.len() == 0 {
            None
        } else {
            Some(initial_params)
        };
        Ok(Models::BinomialLogisticRegression {
            alpha: t.alpha.into(),
            gradient_tolerance: t.gradient_tolerance.into(),
            fit_intercept: t.fit_intercept,
            max_iterations: t.max_iterations,
            initial_params,
        })
    } else if let Some(t) = trainer.multinomial_logistic_regression {
        let initial_params = t
            .initial_params
            .iter()
            .map(|v| (*v).into())
            .collect::<Vec<f64>>();

        let initial_params = if initial_params.len() == 0 {
            None
        } else {
            Some(initial_params)
        };
        Ok(Models::MultinomialLogisticRegression {
            alpha: t.alpha.into(),
            gradient_tolerance: t.gradient_tolerance.into(),
            fit_intercept: t.fit_intercept,
            max_iterations: t.max_iterations,
            initial_params,
            shape: Some(t.shape),
        })
    } else if let Some(t) = trainer.decision_tree {
        {
            let split_quality = match t.split_quality {
                Some(v) => match v {
                    decision_tree::SplitQuality::Gini(decision_tree::Gini {}) => SplitQuality::Gini,
                    decision_tree::SplitQuality::Entropy(decision_tree::Entropy {}) => {
                        SplitQuality::Entropy
                    }
                },
                None => {
                    return Err(Status::failed_precondition("SplitQuality not found!"));
                }
            };

            let max_depth: Option<usize> = match t.max_depth {
                Some(v) => Some(v as usize),
                None => {
                    return Err(Status::failed_precondition("max_depth not provided!"));
                }
            };

            let min_impurity_decrease = t.min_impurity_decrease.into();

            Ok(Models::DecisionTree {
                split_quality,
                max_depth,
                min_weight_split: t.min_weight_split,
                min_weight_leaf: t.min_weight_leaf,
                min_impurity_decrease,
            })
        }
    } else if let Some(t) = trainer.elastic_net {
        let BastionElasticNet {
            penalty,
            l1_ratio,
            with_intercept,
            max_iterations,
            tolerance,
        } = t;
        Ok(Models::ElasticNet {
            penalty,
            l1_ratio,
            with_intercept,
            max_iterations,
            tolerance,
        })
    } else if let Some(t) = trainer.kmeans {
        let BastionKMeans {
            n_runs,
            n_clusters,
            tolerance,
            max_n_iterations,
            init_method,
        } = t;
        let init_method =
            init_method.ok_or_else(|| Status::aborted("Invalid KMeans Init Method!"))?;
        let init_method: KMeansInit<f64> = match init_method {
            k_means::InitMethod::Random(k_means::Random {}) => KMeansInit::Random,
            k_means::InitMethod::PreComputed(k_means::Precomputed {
                list,
                n_centroids,
                n_features,
            }) => {
                let sh = (n_centroids as usize, n_features as usize);
                let list = to_status_error(Array2::from_shape_vec(
                    sh,
                    list.into_iter().map(|v| v.into()).collect::<Vec<_>>(),
                ))?;

                KMeansInit::Precomputed(list)
            }
            k_means::InitMethod::KmeansPara(k_means::KMeansPara {}) => KMeansInit::KMeansPara,
            k_means::InitMethod::KmeansPlusPlus(k_means::KMeansPlusPlus {}) => {
                KMeansInit::KMeansPlusPlus
            }
        };
        Ok(Models::KMeans {
            n_runs: n_runs.try_into().unwrap(),
            n_clusters: n_clusters.try_into().unwrap(),
            tolerance: tolerance.into(),
            max_n_iterations,
            init_method,
        })
    } else if let Some(t) = trainer.svm {
        let BastionSvm {
            c,
            eps,
            nu,
            shrinking,
            platt_params,
            kernel_params,
        } = t;

        let platt_params: PlattParams<f64, ()> = match platt_params {
            Some(v) => PlattParams::default()
                .maxiter(v.maxiter as usize)
                .sigma(v.sigma as f64)
                .minstep(v.ministep as f64),
            None => {
                return Err(Status::aborted(
                    "Could not deserialize PlattParams in Svm".to_string(),
                ));
            }
        };

        let kernel_params = match &kernel_params {
            Some(v) => {
                let kernel_type = match &v.kernel_type {
                    Some(t) => match t {
                        svm::kernel_params::KernelType::Dense(_) => KernelType::Dense,
                        svm::kernel_params::KernelType::Sparse(s) => {
                            KernelType::Sparse(s.sparsity as usize)
                        }
                    },
                    None => {
                        return Err(Status::aborted(
                            "Could not deserialize KernelType in KernelParams in Svm".to_string(),
                        ));
                    }
                };

                // Verify this.
                let n = match v.n() {
                    svm::kernel_params::N::LinearSearch => CommonNearestNeighbour::LinearSearch,
                    svm::kernel_params::N::KdTree => CommonNearestNeighbour::KdTree,
                    svm::kernel_params::N::BallTree => CommonNearestNeighbour::BallTree,
                };

                let kernel_method = match &v.kernel_method {
                    Some(m) => match m {
                        svm::kernel_params::KernelMethod::Guassian(g) => {
                            KernelMethod::Gaussian(g.eps as f64)
                        }
                        svm::kernel_params::KernelMethod::Linear(_) => KernelMethod::Linear,
                        svm::kernel_params::KernelMethod::Poly(p) => {
                            KernelMethod::Polynomial(p.constant as f64, p.degree as f64)
                        }
                    },
                    None => {
                        return Err(Status::aborted(
                            "Could not deserialize KernelMethod in KernelParams in Svm".to_string(),
                        ));
                    }
                };

                Kernel::params()
                    .kind(kernel_type)
                    .method(kernel_method)
                    .nn_algo(n)
            }
            None => {
                return Err(Status::aborted(
                    "Could not deserialize KernelParams in Svm".to_string(),
                ));
            }
        };
        Ok(Models::SVM {
            c,
            eps,
            nu,
            shrinking,
            platt_params,
            kernel_params,
        })
    } else if let Some(t) = trainer.tweedie {
        let link = t.link.ok_or(Status::failed_precondition(
            "Please provide a Link in TweedieRegressor: {Log, Logit, Identity}",
        ))?;

        let link = match link {
            ProtoTweedieLink::Identity(_) => Link::Identity,
            ProtoTweedieLink::Log(_) => Link::Log,
            ProtoTweedieLink::Logit(_) => Link::Logit,
        };
        let BastionTweedieRegressor {
            alpha,
            fit_intercept,
            max_iter,
            tol,
            power,
            ..
        } = t;

        Ok(Models::TweedieRegressor {
            fit_intercept,
            alpha: alpha as f64,
            max_iter,
            link,
            tol: tol as f64,
            power: power as f64,
        })
    } else {
        return Err(Status::not_found(format!("Unknown trainer: {:?}", trainer)));
    }
}
